<a href="https://colab.research.google.com/github/Dianaj30/Predicciones/blob/main/Proyecto1_Parte2_Diana_Cotrina.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

PROYECTO:
Predicción de ventas para productos alimenticios que se venden en distintas tiendas.

## SECCIÓN 1: CARGADO DE DATOS##

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
df=pd.read_csv("/content/drive/MyDrive/CODING DOJO-BOOTCAMP DATA ANALYTICS/Ejercicios plataforma Coding/sales_predictions.csv")
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

Cambio de nombre a las columnas para poder visualizar y manipular mejor los datos

In [3]:
df.rename(columns = {'Item_Identifier' : 'item_id',
                   'Item_Weight' : 'item_w',
                   'Item_Fat_Content' : 'item_fat',
                   'Item_Visibility' : 'item_visib',
                   'Item_Type' : 'item_type',
                   'Item_MRP' : 'intem_mrp',
                   'Outlet_Identifier' : 'outlet_id',
                   'Outlet_Establishment_Year' : 'outlet_establish_year',
                   'Outlet_Size' : 'outlet_size',
                   'Outlet_Location_Type' : 'oulet_location_type',
                   'Outlet_Type' : 'outlet_type',
                   'Item_Outlet_Sales' : 'item_outlet_sales'},
          inplace = True)
df.head()

item_id  item_w item_fat  item_visib              item_type  intem_mrp  \
0   FDA15    9.30  Low Fat    0.016047                  Dairy   249.8092   
1   DRC01    5.92  Regular    0.019278            Soft Drinks    48.2692   
2   FDN15   17.50  Low Fat    0.016760                   Meat   141.6180   
3   FDX07   19.20  Regular    0.000000  Fruits and Vegetables   182.0950   
4   NCD19    8.93  Low Fat    0.000000              Household    53.8614   

  outlet_id  outlet_establish_year outlet_size oulet_location_type  \
0    OUT049                   1999      Medium              Tier 1   
1    OUT018                   2009      Medium              Tier 3   
2    OUT049                   1999      Medium              Tier 1   
3    OUT010                   1998         NaN              Tier 3   
4    OUT013                   1987        High              Tier 3   

         outlet_type  item_outlet_sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

##SECCIÓN 2: EXPLORACIÓN DE DATOS##

1. Conocer los valores y tipos de datos

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   item_id                8523 non-null   object 
 1   item_w                 7060 non-null   float64
 2   item_fat               8523 non-null   object 
 3   item_visib             8523 non-null   float64
 4   item_type              8523 non-null   object 
 5   intem_mrp              8523 non-null   float64
 6   outlet_id              8523 non-null   object 
 7   outlet_establish_year  8523 non-null   int64  
 8   outlet_size            6113 non-null   object 
 9   oulet_location_type    8523 non-null   object 
 10  outlet_type            8523 non-null   object 
 11  item_outlet_sales      8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Se observa que faltan valores en la columna item_w y oulet size

2. Conocer los valores nulos

In [5]:
df.isnull().sum()

item_id                     0
item_w                   1463
item_fat                    0
item_visib                  0
item_type                   0
intem_mrp                   0
outlet_id                   0
outlet_establish_year       0
outlet_size              2410
oulet_location_type         0
outlet_type                 0
item_outlet_sales           0
dtype: int64

Faltan 1463 valores en la columna item _w y 2410 en la columna *outlet_size*

3. Características descriptivas para ver como se comportan los datos

In [6]:
df.describe().T

count         mean          std       min  \
item_w                 7060.0    12.857645     4.643456     4.555   
item_visib             8523.0     0.066132     0.051598     0.000   
intem_mrp              8523.0   140.992782    62.275067    31.290   
outlet_establish_year  8523.0  1997.831867     8.371760  1985.000   
item_outlet_sales      8523.0  2181.288914  1706.499616    33.290   

                               25%          50%          75%           max  
item_w                    8.773750    12.600000    16.850000     21.350000  
item_visib                0.026989     0.053931     0.094585      0.328391  
intem_mrp                93.826500   143.012800   185.643700    266.888400  
outlet_establish_year  1987.000000  1999.000000  2004.000000   2009.000000  
item_outlet_sales       834.247400  1794.331000  3101.296400  13086.964800

2. Revisión 'item_id'

In [7]:
df['item_id'].value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: item_id, Length: 1559, dtype: int64

Se observa que hay 1559 valores únicos

3. Revisión 'item_w'

In [8]:
df['item_w'].value_counts()

12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.275      2
7.685      1
9.420      1
6.520      1
5.400      1
Name: item_w, Length: 415, dtype: int64

In [9]:
df['item_w'].isnull().sum()

1463

Como se había indicado anteriormente existen 1463 valores nulos en el 'item_w'. Sin embargo, como existe la variable 'item_id' que equivale a la identificación del producto, eso nos lleva a la conclusión que por cada id ( cada producto) hay un único peso. Asi que se imputarían los valores de los pesos faltantes de acuerdo al id del producto.

##IMPUTACIÓN DE VALORES A LOS NULOS##

Comprobar la hipóstesis anterior. Ver cabecera de las columnas item_id e item_w. y ordenarlo en base a la columna item_id

In [10]:
df[['item_id','item_w']].sort_values(by='item_id').head(30)

item_id  item_w
8043   DRA12  11.600
118    DRA12  11.600
1693   DRA12  11.600
1197   DRA12  11.600
1245   DRA12  11.600
7467   DRA12  11.600
8195   DRA24  19.350
6863   DRA24  19.350
2879   DRA24     NaN
4416   DRA24     NaN
4900   DRA24  19.350
4130   DRA24  19.350
1148   DRA24  19.350
3927   DRA59   8.270
1876   DRA59   8.270
6338   DRA59   8.270
3962   DRA59   8.270
8371   DRA59   8.270
6179   DRA59     NaN
6057   DRA59     NaN
5609   DRA59   8.270
3465   DRB01   7.390
3733   DRB01     NaN
3571   DRB01   7.390
1836   DRB13   6.115
2479   DRB13   6.115
6626   DRB13   6.115
4135   DRB13   6.115
1431   DRB13   6.115
7811   DRB24   8.785

Se observa que por cada ID del producto corresponde un solo peso y en algunos casos hay valores faltantes que podemos completar usando la tabla de id.

##APLICACION MERCE##

1. Generar dataframe para merge (para merge se debe tener dos tablas, por esa se está generando un nuevo dataframe)

In [11]:
df_w=df.groupby(['item_id'])[['item_w']].max().reset_index()
df_w.rename(columns = {'item_w' : 'item_w_new'},inplace = True)
df_w

item_id  item_w_new
0      DRA12      11.600
1      DRA24      19.350
2      DRA59       8.270
3      DRB01       7.390
4      DRB13       6.115
...      ...         ...
1554   NCZ30       6.590
1555   NCZ41      19.850
1556   NCZ42      10.500
1557   NCZ53       9.600
1558   NCZ54      14.650

[1559 rows x 2 columns]

2. Desarrollo del Merge y la tabla está ordenada según 'item_id'

In [12]:
df_new = pd.merge(left=df, right=df_w, left_on='item_id', right_on='item_id', how='inner')
df_new.sort_values(by='item_id').head(20)

item_id  item_w item_fat  item_visib    item_type  intem_mrp outlet_id  \
680    DRA12   11.60  Low Fat    0.041113  Soft Drinks   142.0154    OUT018   
678    DRA12   11.60  Low Fat    0.040912  Soft Drinks   142.3154    OUT013   
677    DRA12   11.60  Low Fat    0.000000  Soft Drinks   141.6154    OUT045   
676    DRA12   11.60  Low Fat    0.041178  Soft Drinks   140.3154    OUT017   
681    DRA12   11.60  Low Fat    0.068535  Soft Drinks   143.0154    OUT010   
679    DRA12   11.60       LF    0.000000  Soft Drinks   141.9154    OUT035   
4801   DRA24   19.35  Regular    0.066832  Soft Drinks   163.8868    OUT010   
4805   DRA24   19.35  Regular    0.039895  Soft Drinks   162.4868    OUT013   
4804   DRA24   19.35  Regular    0.039990  Soft Drinks   165.0868    OUT049   
4800   DRA24     NaN  Regular    0.069909  Soft Drinks   163.2868    OUT019   
4799   DRA24   19.35  Regular    0.040154  Soft Drinks   164.6868    OUT017   
4802   DRA24     NaN  Regular    0.039735  Soft Drinks   165.7868    OUT027   
4803   DRA24   19.35  Regular    0.039921  Soft Drinks   163.3868    OUT035   
6346   DRA59    8.27  Regular    0.214125  Soft Drinks   183.9924    OUT010   
6344   DRA59     NaN  Regular    0.223985  Soft Drinks   186.2924    OUT019   
6343   DRA59     NaN  Regular    0.127308  Soft Drinks   186.6924    OUT027   
6342   DRA59    8.27  Regular    0.000000  Soft Drinks   183.2924    OUT017   
6341   DRA59    8.27  Regular    0.127821  Soft Drinks   185.9924    OUT013   
6340   DRA59    8.27  Regular    0.128127  Soft Drinks   183.6924    OUT049   
6339   DRA59    8.27  Regular    0.127928  Soft Drinks   184.8924    OUT046   

      outlet_establish_year outlet_size oulet_location_type  \
680                    2009      Medium              Tier 3   
678                    1987        High              Tier 3   
677                    2002         NaN              Tier 2   
676                    2007         NaN              Tier 2   
681                    1998         NaN              Tier 3   
679                    2004       Small              Tier 2   
4801                   1998         NaN              Tier 3   
4805                   1987        High              Tier 3   
4804                   1999      Medium              Tier 1   
4800                   1985       Small              Tier 1   
4799                   2007         NaN              Tier 2   
4802                   1985      Medium              Tier 3   
4803                   2004       Small              Tier 2   
6346                   1998         NaN              Tier 3   
6344                   1985       Small              Tier 1   
6343                   1985      Medium              Tier 3   
6342                   2007         NaN              Tier 2   
6341                   1987        High              Tier 3   
6340                   1999      Medium              Tier 1   
6339                   1997       Small              Tier 1   

            outlet_type  item_outlet_sales  item_w_new  
680   Supermarket Type2           850.8924       11.60  
678   Supermarket Type1          2552.6772       11.60  
677   Supermarket Type1          3829.0158       11.60  
676   Supermarket Type1          2552.6772       11.60  
681       Grocery Store           283.6308       11.60  
679   Supermarket Type1           992.7078       11.60  
4801      Grocery Store           327.5736       19.35  
4805  Supermarket Type1          4422.2436       19.35  
4804  Supermarket Type1           982.7208       19.35  
4800      Grocery Store           491.3604       19.35  
4799  Supermarket Type1          1146.5076       19.35  
4802  Supermarket Type3          4913.6040       19.35  
4803  Supermarket Type1          3439.5228       19.35  
6346      Grocery Store           185.0924        8.27  
6344      Grocery Store           555.2772        8.27  
6343  Supermarket Type3          7033.5112        8.27  
6342  Supermarket Type1          2406.2012        8.

In [13]:
df_new.isnull().sum()

item_id                     0
item_w                   1463
item_fat                    0
item_visib                  0
item_type                   0
intem_mrp                   0
outlet_id                   0
outlet_establish_year       0
outlet_size              2410
oulet_location_type         0
outlet_type                 0
item_outlet_sales           0
item_w_new                  4
dtype: int64

In [14]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8523 entries, 0 to 8522
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   item_id                8523 non-null   object 
 1   item_w                 7060 non-null   float64
 2   item_fat               8523 non-null   object 
 3   item_visib             8523 non-null   float64
 4   item_type              8523 non-null   object 
 5   intem_mrp              8523 non-null   float64
 6   outlet_id              8523 non-null   object 
 7   outlet_establish_year  8523 non-null   int64  
 8   outlet_size            6113 non-null   object 
 9   oulet_location_type    8523 non-null   object 
 10  outlet_type            8523 non-null   object 
 11  item_outlet_sales      8523 non-null   float64
 12  item_w_new             8519 non-null   float64
dtypes: float64(5), int64(1), object(7)
memory usage: 932.2+ KB


Se observa que los valores nulos se redujeron a 4 nada más. Para poder saber si se eliminan o no, se hará el porcentaje para ver cuanto representa del valor total

Analizando los valores perdidos

In [15]:
df_new.isna().sum()/df.shape[0]*100

item_id                   0.000000
item_w                   17.165317
item_fat                  0.000000
item_visib                0.000000
item_type                 0.000000
intem_mrp                 0.000000
outlet_id                 0.000000
outlet_establish_year     0.000000
outlet_size              28.276428
oulet_location_type       0.000000
outlet_type               0.000000
item_outlet_sales         0.000000
item_w_new                0.046932
dtype: float64

Los items nulos representan el 0.04%, menos del 10% por lo que se pueden eliminar.

In [16]:
df_new_nulos=df_new[df_new['item_w_new'].isna()]
df_new_nulos

item_id  item_w item_fat  item_visib     item_type  intem_mrp outlet_id  \
4125   FDN52     NaN  Regular    0.130933  Frozen Foods    86.9198    OUT027   
6440   FDK57     NaN  Low Fat    0.079904   Snack Foods   120.0440    OUT027   
8278   FDE52     NaN  Regular    0.029742         Dairy    88.9514    OUT027   
8463   FDQ60     NaN  Regular    0.191501  Baking Goods   121.2098    OUT019   

      outlet_establish_year outlet_size oulet_location_type  \
4125                   1985      Medium              Tier 3   
6440                   1985      Medium              Tier 3   
8278                   1985      Medium              Tier 3   
8463                   1985       Small              Tier 1   

            outlet_type  item_outlet_sales  item_w_new  
4125  Supermarket Type3          1569.9564         NaN  
6440  Supermarket Type3          4434.2280         NaN  
8278  Supermarket Type3          3453.5046         NaN  
8463      Grocery Store           120.5098         NaN

Eliminación de valores nulos en df_new

In [17]:
df_new.drop([4125,6440,8278,8463], axis=0, inplace=True)
df_new

item_id  item_w item_fat  item_visib           item_type  intem_mrp  \
0      FDA15   9.300  Low Fat    0.016047               Dairy   249.8092   
1      FDA15   9.300  Low Fat    0.016055               Dairy   250.2092   
2      FDA15   9.300  Low Fat    0.016019               Dairy   248.5092   
3      FDA15   9.300  Low Fat    0.016088               Dairy   249.6092   
4      FDA15   9.300  Low Fat    0.026818               Dairy   248.9092   
...      ...     ...      ...         ...                 ...        ...   
8518   NCF55   6.675       LF    0.021666           Household    33.3874   
8519   NCW30   5.210  Low Fat    0.011072           Household   257.8962   
8520   NCW30   5.210  Low Fat    0.011008           Household   259.5962   
8521   NCW05  20.250  Low Fat    0.148303  Health and Hygiene   108.3938   
8522   NCW05  20.250  Low Fat    0.148373  Health and Hygiene   108.8938   

     outlet_id  outlet_establish_year outlet_size oulet_location_type  \
0       OUT049                   1999      Medium              Tier 1   
1       OUT045                   2002         NaN              Tier 2   
2       OUT035                   2004       Small              Tier 2   
3       OUT018                   2009      Medium              Tier 3   
4       OUT010                   1998         NaN              Tier 3   
...        ...                    ...         ...                 ...   
8518    OUT046                   1997       Small              Tier 1   
8519    OUT017                   2007         NaN              Tier 2   
8520    OUT035                   2004       Small              Tier 2   
8521    OUT049                   1999      Medium              Tier 1   
8522    OUT045                   2002         NaN              Tier 2   

            outlet_type  item_outlet_sales  item_w_new  
0     Supermarket Type1          3735.1380       9.300  
1     Supermarket Type1          5976.2208       9.300  
2     Supermarket Type1          6474.2392       9.300  
3     Supermarket Type2          5976.2208       9.300  
4         Grocery Store           498.0184       9.300  
...                 ...                ...         ...  
8518  Supermarket Type1          1235.0590       6.675  
8519  Supermarket Type1          4402.9354       5.210  
8520  Supermarket Type1          2848.9582       5.210  
8521  Supermarket Type1          2787.0388      20.250  
8522  Supermarket Type1          3644.5892      20.250  

[8519 rows x 13 columns]

In [18]:
df_new.isnull().sum()

item_id                     0
item_w                   1459
item_fat                    0
item_visib                  0
item_type                   0
intem_mrp                   0
outlet_id                   0
outlet_establish_year       0
outlet_size              2410
oulet_location_type         0
outlet_type                 0
item_outlet_sales           0
item_w_new                  0
dtype: int64

In [19]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8519 entries, 0 to 8522
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   item_id                8519 non-null   object 
 1   item_w                 7060 non-null   float64
 2   item_fat               8519 non-null   object 
 3   item_visib             8519 non-null   float64
 4   item_type              8519 non-null   object 
 5   intem_mrp              8519 non-null   float64
 6   outlet_id              8519 non-null   object 
 7   outlet_establish_year  8519 non-null   int64  
 8   outlet_size            6109 non-null   object 
 9   oulet_location_type    8519 non-null   object 
 10  outlet_type            8519 non-null   object 
 11  item_outlet_sales      8519 non-null   float64
 12  item_w_new             8519 non-null   float64
dtypes: float64(5), int64(1), object(7)
memory usage: 931.8+ KB


##Revisión 'outlet_size'##





1. Conocer los valores únicos

In [20]:
df_new['outlet_size'].value_counts()

Medium    2790
Small     2387
High       932
Name: outlet_size, dtype: int64

Existen tres tipos de tiendas "Medium", "Small", "High"

2. Conocer los valores nulos

In [21]:
df_new['outlet_size'].isnull().sum()

2410

Existen 2410 valores pedidos en el tamaño de tienda

3. Búsqueda de valores perdidos, poder saber dónde se encuentran

In [22]:
outlet_size_nulos=df[df['outlet_size'].isna()]
outlet_size_nulos

item_id  item_w item_fat  item_visib              item_type  intem_mrp  \
3      FDX07  19.200  Regular    0.000000  Fruits and Vegetables   182.0950   
8      FDH17  16.200  Regular    0.016687           Frozen Foods    96.9726   
9      FDU28  19.200  Regular    0.094450           Frozen Foods   187.8214   
25     NCD06  13.000  Low Fat    0.099887              Household    45.9060   
28     FDE51   5.925  Regular    0.161467                  Dairy    45.5086   
...      ...     ...      ...         ...                    ...        ...   
8502   NCH43   8.420  Low Fat    0.070712              Household   216.4192   
8508   FDW31  11.350  Regular    0.043246  Fruits and Vegetables   199.4742   
8509   FDG45   8.100  Low Fat    0.214306  Fruits and Vegetables   213.9902   
8514   FDA01  15.000  Regular    0.054489                 Canned    57.5904   
8519   FDS36   8.380  Regular    0.046982           Baking Goods   108.1570   

     outlet_id  outlet_establish_year outlet_size oulet_location_type  \
3       OUT010                   1998         NaN              Tier 3   
8       OUT045                   2002         NaN              Tier 2   
9       OUT017                   2007         NaN              Tier 2   
25      OUT017                   2007         NaN              Tier 2   
28      OUT010                   1998         NaN              Tier 3   
...        ...                    ...         ...                 ...   
8502    OUT045                   2002         NaN              Tier 2   
8508    OUT045                   2002         NaN              Tier 2   
8509    OUT010                   1998         NaN              Tier 3   
8514    OUT045                   2002         NaN              Tier 2   
8519    OUT045                   2002         NaN              Tier 2   

            outlet_type  item_outlet_sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

Se observa que hay una relación entre los tipos de tienda y los valores nulos, por ello se los agrupará para observar a que tamaño de tiendan pertenecen cada uno de ellos.

4. Realización del groupby

In [23]:
df_new_size=df_new.groupby(['outlet_type', 'oulet_location_type','outlet_size'])[['item_outlet_sales']].mean()
df_new_size

item_outlet_sales
outlet_type       oulet_location_type outlet_size                   
Grocery Store     Tier 1              Small               340.746838
Supermarket Type1 Tier 1              Medium             2348.354635
                                      Small              2277.844267
                  Tier 2              Small              2438.841866
                  Tier 3              High               2298.995256
Supermarket Type2 Tier 3              Medium             1995.498739
Supermarket Type3 Tier 3              Medium             3695.781505

En base a este análisis se observa lo siguiente:
1. Grocery store Tier1 solo tiene el tamaño 'Small"
2. Supermarket Type1 Tier2 su tamaño es 'Small'
4. Supermarket Type1 Tier3 su tamaño es 'High'
5. Supermarket Type2 Tier3 su tamaño es 'Midium'
6. Supermarket Type3 Tier3 su tamaño es 'Medium'

5. Realización de iteracion con for para reemplazar los valores

In [24]:
for index,row in df_new.iterrows():
  if(row['outlet_type']=='Grocery Store') and (row['oulet_location_type']=='Tier 1'):
    df_new.at[index,"outlet_size_new"] = 'Small'
  elif (row['outlet_type']=='Supermarket Type1') and (row['oulet_location_type']=='Tier 2'):
    df_new.at[index,"outlet_size_new"] ='Small'
  elif (row['outlet_type']=='Supermarket Type1') and (row['oulet_location_type']=='Tier 3'):
    df_new.at[index,"outlet_size_new"] ='High'
  elif (row['outlet_type']=='Supermarket Type2') and (row['oulet_location_type']=='Tier 3'):
    df_new.at[index,"outlet_size_new"] ='Medium'
  elif (row['outlet_type']=='Supermarket Type3') and (row['oulet_location_type']=='Tier 3'):
    df_new.at[index,"outlet_size_new"] ='Medium'
  else:
    df_new.at[index,"outlet_size_new"] = row['outlet_size']

6. Revisión de valores nulos

In [25]:
df_new['outlet_size'].isnull().sum()

2410

In [26]:
df_new['outlet_size_new'].isnull().sum()

555

In [27]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8519 entries, 0 to 8522
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   item_id                8519 non-null   object 
 1   item_w                 7060 non-null   float64
 2   item_fat               8519 non-null   object 
 3   item_visib             8519 non-null   float64
 4   item_type              8519 non-null   object 
 5   intem_mrp              8519 non-null   float64
 6   outlet_id              8519 non-null   object 
 7   outlet_establish_year  8519 non-null   int64  
 8   outlet_size            6109 non-null   object 
 9   oulet_location_type    8519 non-null   object 
 10  outlet_type            8519 non-null   object 
 11  item_outlet_sales      8519 non-null   float64
 12  item_w_new             8519 non-null   float64
 13  outlet_size_new        7964 non-null   object 
dtypes: float64(5), int64(1), object(8)
memory usage: 1.2+ MB

Se observa que ahora hay 555 valores perdidos, ya no 2410.

7. Búsqueda de valores nulos

In [28]:
outlet_size_new_nulos=df_new[df_new['outlet_size_new'].isna()]
outlet_size_new_nulos

item_id  item_w item_fat  item_visib              item_type  intem_mrp  \
4      FDA15   9.300  Low Fat    0.026818                  Dairy   248.9092   
15     FDN15  17.500  Low Fat    0.028009                   Meat   141.7180   
21     FDX07  19.200  Regular    0.000000  Fruits and Vegetables   182.0950   
28     NCD19   8.930  Low Fat    0.022060              Household    56.0614   
38     FDO10  13.650  Regular    0.021344            Snack Foods    58.3588   
...      ...     ...      ...         ...                    ...        ...   
8394   NCP17  19.350  Low Fat    0.000000     Health and Hygiene    65.3168   
8462   NCA18  10.100  Low Fat    0.093862              Household   115.9492   
8478   FDP15  15.200  Low Fat    0.000000                   Meat   256.0330   
8482   NCD55  14.000  Low Fat    0.040725              Household    43.4454   
8506   NCU53   5.485  Low Fat    0.071558     Health and Hygiene   165.7842   

     outlet_id  outlet_establish_year outlet_size oulet_location_type  \
4       OUT010                   1998         NaN              Tier 3   
15      OUT010                   1998         NaN              Tier 3   
21      OUT010                   1998         NaN              Tier 3   
28      OUT010                   1998         NaN              Tier 3   
38      OUT010                   1998         NaN              Tier 3   
...        ...                    ...         ...                 ...   
8394    OUT010                   1998         NaN              Tier 3   
8462    OUT010                   1998         NaN              Tier 3   
8478    OUT010                   1998         NaN              Tier 3   
8482    OUT010                   1998         NaN              Tier 3   
8506    OUT010                   1998         NaN              Tier 3   

        outlet_type  item_outlet_sales  item_w_new outlet_size_new  
4     Grocery Store           498.0184       9.300             NaN  
15    Grocery Store           419.4540      17.500             NaN  
21    Grocery Store           732.3800      19.200             NaN  
28    Grocery Store           276.3070       8.930             NaN  
38    Grocery Store           114.5176      13.650             NaN  
...             ...                ...         ...             ...  
8394  Grocery Store           191.7504      19.350             NaN  
8462  Grocery Store           231.6984      10.100             NaN  
8478  Grocery Store          1281.6650      15.200             NaN  
8482  Grocery Store            41.9454      14.000             NaN  
8506  Grocery Store           165.7842       5.485             NaN  

[555 rows x 14 columns]

In [29]:
df_new['outlet_id'].value_counts()

OUT013    932
OUT027    932
OUT049    930
OUT035    930
OUT046    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    527
Name: outlet_id, dtype: int64

El supuesto es que el 'outlet_id' OUT010 tiene los 555 valores faltantes, es decir en ese outlet nunca se registró el tamaño.

8. Para comprobar la hipótesis se hará una búsqueda del tamaño de tienda por 'outlet_id'

In [30]:
df_new_id_outlet=df_new.groupby(['outlet_id','outlet_type','oulet_location_type'])[['outlet_size_new']].count()
df_new_id_outlet

outlet_size_new
outlet_id outlet_type       oulet_location_type                 
OUT010    Grocery Store     Tier 3                             0
OUT013    Supermarket Type1 Tier 3                           932
OUT017    Supermarket Type1 Tier 2                           926
OUT018    Supermarket Type2 Tier 3                           928
OUT019    Grocery Store     Tier 1                           527
OUT027    Supermarket Type3 Tier 3                           932
OUT035    Supermarket Type1 Tier 2                           930
OUT045    Supermarket Type1 Tier 2                           929
OUT046    Supermarket Type1 Tier 1                           930
OUT049    Supermarket Type1 Tier 1                           930

Acá comprobamos la hipótesis que los valores faltantes corresponden a la tienda con id OUT010

Debido a que es una tienda completa no correspondería eliminarla, porque se requiere analizar el nivel de predicción de ventas de las tiendas.

Ahora hacemos un análisis del promedio de ventas por tienda y es bastante variado, entonces no se elimina.

In [31]:
df_new_id_outlet_sales=df_new.groupby(['outlet_id','outlet_type','oulet_location_type'])[['item_outlet_sales']].mean()
df_new_id_outlet_sales

item_outlet_sales
outlet_id outlet_type       oulet_location_type                   
OUT010    Grocery Store     Tier 3                      339.351662
OUT013    Supermarket Type1 Tier 3                     2298.995256
OUT017    Supermarket Type1 Tier 2                     2340.675263
OUT018    Supermarket Type2 Tier 3                     1995.498739
OUT019    Grocery Store     Tier 1                      340.746838
OUT027    Supermarket Type3 Tier 3                     3695.781505
OUT035    Supermarket Type1 Tier 2                     2438.841866
OUT045    Supermarket Type1 Tier 2                     2192.384798
OUT046    Supermarket Type1 Tier 1                     2277.844267
OUT049    Supermarket Type1 Tier 1                     2348.354635

Ahora analizamos el promedio de ventas por tamaño de tienda para ver si se puede eliminar la columna de tamaño de tienda

In [32]:
df_new_size_sales=df_new.groupby(['outlet_size_new'])[['item_outlet_sales']].mean()
df_new_size_sales

item_outlet_sales
outlet_size_new                   
High                   2298.995256
Medium                 2681.097134
Small                  2067.487614

Se observa que no hay mucha variación entre el tamaño de tienda

Para comprobar la hipótesis se verá la desviación estándar

In [33]:
df_new_size_desv= {'outlet_size': ['High', 'Medium', 'Small'],
        'item_outlet_sales': [10256.6490, 13086.9648, 9779.9362]}
df_new_size_desv = pd.DataFrame(df_new_size_desv)

In [34]:
df_new_size_desv.describe()

item_outlet_sales
count           3.000000
mean        11041.183333
std          1787.660519
min          9779.936200
25%         10018.292600
50%         10256.649000
75%         11671.806900
max         13086.964800

Se observa que la desviación estándar si es alta, se buscará otras comparaciones para ver si se puede o no eliminar la tamaño de tienda.

Ahora vamos a revisar si hay variación por id, localización y tipo. Se observa que la variación es mucho mayor en comparación con la variación del tamaño de tienda.

In [35]:
df_new_sales_id_outlet=df_new.groupby(['outlet_id','outlet_type','oulet_location_type'])[['item_outlet_sales']].mean()
df_new_sales_id_outlet

item_outlet_sales
outlet_id outlet_type       oulet_location_type                   
OUT010    Grocery Store     Tier 3                      339.351662
OUT013    Supermarket Type1 Tier 3                     2298.995256
OUT017    Supermarket Type1 Tier 2                     2340.675263
OUT018    Supermarket Type2 Tier 3                     1995.498739
OUT019    Grocery Store     Tier 1                      340.746838
OUT027    Supermarket Type3 Tier 3                     3695.781505
OUT035    Supermarket Type1 Tier 2                     2438.841866
OUT045    Supermarket Type1 Tier 2                     2192.384798
OUT046    Supermarket Type1 Tier 1                     2277.844267
OUT049    Supermarket Type1 Tier 1                     2348.354635

Al analizar el promedio de ventas por tipo de tienda y localización, se observa que el tipo de tienda "Grovery Store" ubicada en Tier3 tiene un promedio de ventas de 339.35. Mientras que tienda la tienda"Grovery Store" ubicada en Tier1 tiene 340.74. La diferencia del promedio de ventas de ambas tiendas es 1.39. Un valor bastante bajo, es decir, no hay tanto variación si comparamos que los otros promedios de ventas.

Anteriormente habiamos observado que la tienda "Grovery Store" ubicada en Tier1 solo tiene una asignación de tamaño que es "Small". Por lo que, se podría asignar el tamaño "Small" a las tiendas "Grocery Store" Tier3 ya que su promedio de ventas es bastante similar.

Asignación de "Small" a los valores faltantes en tipo de tiendas por la razon antes expuesta

In [36]:
df_new['outlet_size_new'].fillna('Small', inplace=True)
df_new

item_id  item_w item_fat  item_visib           item_type  intem_mrp  \
0      FDA15   9.300  Low Fat    0.016047               Dairy   249.8092   
1      FDA15   9.300  Low Fat    0.016055               Dairy   250.2092   
2      FDA15   9.300  Low Fat    0.016019               Dairy   248.5092   
3      FDA15   9.300  Low Fat    0.016088               Dairy   249.6092   
4      FDA15   9.300  Low Fat    0.026818               Dairy   248.9092   
...      ...     ...      ...         ...                 ...        ...   
8518   NCF55   6.675       LF    0.021666           Household    33.3874   
8519   NCW30   5.210  Low Fat    0.011072           Household   257.8962   
8520   NCW30   5.210  Low Fat    0.011008           Household   259.5962   
8521   NCW05  20.250  Low Fat    0.148303  Health and Hygiene   108.3938   
8522   NCW05  20.250  Low Fat    0.148373  Health and Hygiene   108.8938   

     outlet_id  outlet_establish_year outlet_size oulet_location_type  \
0       OUT049                   1999      Medium              Tier 1   
1       OUT045                   2002         NaN              Tier 2   
2       OUT035                   2004       Small              Tier 2   
3       OUT018                   2009      Medium              Tier 3   
4       OUT010                   1998         NaN              Tier 3   
...        ...                    ...         ...                 ...   
8518    OUT046                   1997       Small              Tier 1   
8519    OUT017                   2007         NaN              Tier 2   
8520    OUT035                   2004       Small              Tier 2   
8521    OUT049                   1999      Medium              Tier 1   
8522    OUT045                   2002         NaN              Tier 2   

            outlet_type  item_outlet_sales  item_w_new outlet_size_new  
0     Supermarket Type1          3735.1380       9.300          Medium  
1     Supermarket Type1          5976.2208       9.300           Small  
2     Supermarket Type1          6474.2392       9.300           Small  
3     Supermarket Type2          5976.2208       9.300          Medium  
4         Grocery Store           498.0184       9.300           Small  
...                 ...                ...         ...             ...  
8518  Supermarket Type1          1235.0590       6.675           Small  
8519  Supermarket Type1          4402.9354       5.210           Small  
8520  Supermarket Type1          2848.9582       5.210           Small  
8521  Supermarket Type1          2787.0388      20.250          Medium  
8522  Supermarket Type1          3644.5892      20.250           Small  

[8519 rows x 14 columns]

In [37]:
df_new_size_sales_new=df_new.groupby(['outlet_size'])[['item_outlet_sales']].mean()
df_new_size_sales_new

item_outlet_sales
outlet_size                   
High               2298.995256
Medium             2681.097134
Small              1912.899743

Nuevo promedio de ventas si se imputa "Small" a la tienda Grocery Store Tier3. El nuevo promedio es 1912.14, el anterior era 2067.028. La diferencia es de 155 aprox. La diferencia es menor en comparación si se eliminaba la columna entera de tamaño de tienda. Donde habia una diferencia de 383 a 614 entre tamaño de tiendas.

In [38]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8519 entries, 0 to 8522
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   item_id                8519 non-null   object 
 1   item_w                 7060 non-null   float64
 2   item_fat               8519 non-null   object 
 3   item_visib             8519 non-null   float64
 4   item_type              8519 non-null   object 
 5   intem_mrp              8519 non-null   float64
 6   outlet_id              8519 non-null   object 
 7   outlet_establish_year  8519 non-null   int64  
 8   outlet_size            6109 non-null   object 
 9   oulet_location_type    8519 non-null   object 
 10  outlet_type            8519 non-null   object 
 11  item_outlet_sales      8519 non-null   float64
 12  item_w_new             8519 non-null   float64
 13  outlet_size_new        8519 non-null   object 
dtypes: float64(5), int64(1), object(8)
memory usage: 1.2+ MB

In [39]:
df_new.isnull().sum()

item_id                     0
item_w                   1459
item_fat                    0
item_visib                  0
item_type                   0
intem_mrp                   0
outlet_id                   0
outlet_establish_year       0
outlet_size              2410
oulet_location_type         0
outlet_type                 0
item_outlet_sales           0
item_w_new                  0
outlet_size_new             0
dtype: int64

En conclusión:
Solo se eliminaron 4 valores, los demás valores nulos se imputaron al hacer comparaciones con otras variables, según el explicado en el procedimiento. Por lo tanto, no queda ningún valor nulo.

##Revisión de la columna "item_fat"##

In [40]:
df_new['item_fat'].value_counts()

Low Fat    5088
Regular    2886
LF          316
reg         117
low fat     112
Name: item_fat, dtype: int64

Hay 5 valores de "item fat" sin embargo se presume que LF, reg son abreviaciones de Low Fat y Regular, por otro lado, low fat es quivalente a Low Fat. Por otro lado, resulta extraño que sean solo cientos mientras que Low Fat y Regular sean mile.

Por ello, se reemplazará sus valores para que sean todos entre Low Fat y Regular.

In [44]:
df_new.replace(['LF'], 'Low Fat', inplace=True)

In [45]:
df_new.replace(['reg'], 'Regular', inplace=True)

In [46]:
df_new.replace(['low fat'], 'Low Fat', inplace=True)

In [47]:
df_new['item_fat'].value_counts()

Low Fat    5516
Regular    3003
Name: item_fat, dtype: int64

Con estos cambios se tendría la base de datos limpia